### ISSUE TITLE
Dex EnvoyFilter forces authentication to existing istio services

### ISSUE THREAD

**Issue by barlock** [kubeflow/kubeflow#4549]
/kind bug

**What steps did you take and what happened:**
Created a cluster (IBM's cloud) configured with istio and knative. 

Created a knative service (behind istio) in the default namespace, confirmed it works at my hostname with an istio ingress created for it. 

Installed kubeflow 0.7.0 configured with dex, and without installing istio (as it was already there). There is a modification here also to limit it to kubeflow.<cluster>.containers.appdomain.cloud to keep other services working.

Confirmed kubeflow works with dex authentication. 

All istio services (specifically, the knative ones) are also redirected to dex and authenticated.

**What did you expect to happen:**
All kubeflow services are authenticated with dex.

All knative services are not authenticated with dex.

**Anything else you would like to add:**
I tried moving the EnvoyFilter from `istio-system` to `kubeflow` and that broke everything. 

The method of knative installation creates the ingress that `kubeflow` also uses. It doesn't appear to have anything knative specific in it, just wires up TLS and loadBalancers.  

**Environment:**

- Kubeflow version: (version number can be found at the bottom left corner of the Kubeflow dashboard): 0.7.0
- kfctl version: (use `kfctl version`): 0.7.0
- Kubernetes platform: (e.g. `minikube`) IBM IKS
- Kubernetes version: (use `kubectl version`): 1.15.6
- OS (e.g. from `/etc/os-release`): Debian

**Comment by yanniszark** [comment]
Hi @barlock, thanks for opening an issue.
So, based on what you said, the issue is that the EnvoyFilter also authenticates requests related to knative and you did not anticipate that.
If that's correct, could you update the issue title to reflect that issue?

I can see 2 solutions here:
1. If you don't want authentication in those routes, you could look into the skip_auth_url option of the authservice. It allows you to whitelist certain path prefixes.
2. If you want authentication, you can provide the session token in the x-auth-token header.

**Comment by barlock** [comment]
Can you point me to docs re: `skip_auth_url` I can't figure out where that lives.

I'm curious about assumptions for how Kubeflow is installed. Is it assumed that Kubeflow isn't installed in a multi-tenant cluster? If so why?

My guess is for pre-release simplicity it makes sense, I wasn't sure if this issue is a feature or a bug TBH.

**Comment by yanniszark** [comment]
@barlock `SKIP_AUTH_URL` is an environment variable in the authservice deployment.
The authservice repo is here along with documentation is here: https://github.com/arrikto/oidc-authservice.

The EnvoyFilter selects Gateways based on the `istio: ingressgateway` selector.
I would expect it to only apply to the Kubeflow gateway, but the selector may be too generic.

Do you mean that you had an Istio Gateway, separate to Kubeflow, and its behaviour changed after you installed Kubeflow?

**Comment by barlock** [comment]
SKIP_AUTH_URI is interesting, I need to whitelist hosts I think rather than paths.

> The EnvoyFilter selects Gateways based on the istio: ingressgateway selector.
Interesting, I'm still learning my way around istio, this probably needs some looking into as a path for me.

> Do you mean that you had an Istio Gateway, separate to Kubeflow, and its behaviour changed after you installed Kubeflow?

Yes, previously to kubeflow's installation, I have a working knative installation that set up an istio ingress from cluster's loadbalancers, that routed traffic via a knative gateway. After kubeflow is installed, that knative gateway now (attempts to)reroutes traffic to dex. Thus, knative services no longer work. I also don't want them to be authenticated.

Looking into those gateways, the knative and kubeflow gateways do have a selector `istio: ingressgateway` on them.

**Comment by barlock** [comment]
> The EnvoyFilter selects Gateways based on the `istio: ingressgateway` selector.
I would expect it to only apply to the Kubeflow gateway, but the selector may be too generic.

Just looked into this. Yes, I believe that's true. It will select all gateways that use the default controller. 

My expectation was that, by default, it would only apply to the kubeflow one, specifically `kubeflow-gateway`.

Digging in further, gateways use the same ingress controllers, `workloadLabels` selects ingress controllers, not gateways, hence the behavior I'm seeing. In order to make it work, the only solution I'm seeing is to deploy a whole new ingress gateway controller specifically for kubeflow, which seems like overkill to me. istio 1.3 seems to have updated the EnvoyFilter CRD, but I don't see anything in there or 1.2 (that we're currently using?), that looks like the filter can be applied in the way I'm looking for. Am I not seeing something (hopefully)?

[Docs](https://archive.istio.io/v1.2/docs/reference/config/networking/v1alpha3/envoy-filter/#EnvoyFilter) for `workloadLabels` as it took me a bit to find it

**Comment by yanniszark** [comment]
@barlock thanks for bringing this up, it is definitely a limitation.
I would expect Istio to have support for different Authentication Policies per Gateway, but they seem to only support workload selector, which selects backend Pods and not Gateway CRs.

This seems to be a limitation of Istio, as per the community's answer here:
https://istio.slack.com/archives/C3TEGNZ7W/p1575568960139700

The solution for now indeed seems to be to use a different gateway backend or work with whitelisted paths.
This seems strange to me as well.

**Comment by barlock** [comment]
@yanniszark Can you invite me into that workspace (barlock@ibm.com)?

**Comment by yanniszark** [comment]
@barlock it's the official istio slack, you can join by following the instructions on this page:
https://istio.io/about/community/join/

**Comment by thetume** [comment]
We ran into this issue as well. We have an existing kubernetes cluster with some custom services. After kubeflow installation (Multi-user, auth-enabled Kubeflow with kfctl_existing_arrikto) everything is routed to Dex. This is definitely problematic issue with multi tenant setup. For now we can work around it with whitelisting paths to SKIP_AUTH_URI, but in the long run that will be hard to maintain. 

**Environment**
Kubeflow version: 0.7.0
kfctl version: (use kfctl version): 0.7.0
Kubernetes platform: AWS EC2 mimicking future on-prem
Kubernetes version: (use kubectl version): 1.15.6
OS (e.g. from /etc/os-release): RHEL 7

**Comment by LeonardAukea** [comment]
I think this issue should not be closed. We are currently facing this issue as well with Kubeflow version: 1.0.0

**Comment by achandras** [comment]
I believe this is still an issue, could it be reopened? I am still running into this with a Kubeflow 1.1.0 deployment using existing Istio services

**Comment by aaron-arellano** [comment]
This is still an issue. OIDC-Authservice needs to take multi-tenancy into account. I tried moving filter to sidecar but JWA does not work properly. Whitelisting hosts is not maintainable.

**Comment by niuqg** [comment]
Add whitelist for seldon-core in kubeflow 1.3.
1. Update configmap oidc-authservice-parameters in  the  istio namespace
`SKIP_AUTH_URI: /dex /seldon/`  

2. Create authorizationPolicy:
```
kind: AuthorizationPolicy
metadata:
  labels:
    app.kubernetes.io/managed-by: seldon-core
  name: seldon
  namespace: seldon
spec:
  action: ALLOW
  rules:
  - {}
  selector:
    matchLabels:
      app.kubernetes.io/managed-by: seldon-core
```

**Comment by ReggieCarey** [comment]
Is there any movement on this problem?  It's affecting me in KF 1.3.0.  After Kubeflow is installed, basic Istio Getting started apps like httpbin no longer work.  Kubeflow folks: Please stop assuming that your use of Istio is the ONLY use of Istio on a kubernetes cluster.  If DEX Authentication is in place for Kubeflow, then it should be in place explicitly and ONLY for Kubeflow, not every other application deployed to the Kubernetes environment.

**Comment by jovan-absci** [comment]
While not a completely clean solution, I've managed to solve this using [a custom authorization policy](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/), a feature available in Istio 1.9 and onward.
I rewrote the auth provider spec from the EnvoyFilter into an extension provider (as per the article). For dex it would look like:
`extensionProviders:
    - name: "dex-auth-provider"
      envoyExtAuthzHttp:
        service: "authservice.istio-system.svc.cluster.local"
        port: "8080" 
        includeHeadersInCheck: ["authorization", "cookie", "x-auth-token"]
        headersToUpstreamOnAllow: ["kubeflow-userid"]`
You can skip the "Deploy the external authorizer" section as we already have dex and auth-service in place.

Then I added an authorization policy for Kf:
`apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: dex-auth
  namespace: istio-system
spec:
  selector:
    matchLabels:
      istio: ingressgateway
  action: CUSTOM
  provider:
    # The provider name must match the extension provider defined in the mesh config.
    name: dex-auth-provider
  rules:
  # The rules specify when to trigger the external authorizer.
  - to:
    - operation:
        hosts: ["kf.example.com"]`
You can match on [different things](https://istio.io/latest/docs/reference/config/security/authorization-policy/#Operation) such as paths, methods, ports etc. It appears that the policies need to reside in the istio-system namespaces, otherwise it doesn't work.

After this is in place, delete the authn_filter EnvoyFilter, restart istiod deployment and it should work.

**Comment by jovan-absci** [comment]
While not a completely clean solution, I've managed to solve this using [a custom authorization policy](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/), a feature available in Istio 1.9 and onward.
I rewrote the auth provider spec from the EnvoyFilter into an extension provider (as per the article). For dex it would look like:
```
extensionProviders:
    - name: dex-auth-provider
      envoyExtAuthzHttp:
        service: "authservice.istio-system.svc.cluster.local"
        port: "8080" 
        includeHeadersInCheck: ["authorization", "cookie", "x-auth-token"]
        headersToUpstreamOnAllow: ["kubeflow-userid"]
```

You can skip the "Deploy the external authorizer" section as we already have dex and auth-service in place.

Then I added an authorization policy for Kf:
```
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: dex-auth
  namespace: istio-system
spec:
  selector:
    matchLabels:
      istio: ingressgateway
  action: CUSTOM
  provider:
    # The provider name must match the extension provider defined in the mesh config.
    name: dex-auth-provider
  rules:
  # The rules specify when to trigger the external authorizer.
  - to:
    - operation:
        hosts: ["kf.example.com"]
```
You can match on [different things](https://istio.io/latest/docs/reference/config/security/authorization-policy/#Operation) such as paths, methods, ports etc. It appears that the policies need to reside in the istio-system namespace, otherwise it doesn't work.

After this is in place, delete the authn_filter EnvoyFilter, restart istiod deployment and it should work.

**Comment by N214** [comment]
Not sure if I did this correctly, add the istio configmap and the AuthorizationPolicy. Also deleted the authn_filter EnvoyFilter, restart istiod deployment. 
The dex redirect did disappear, and I see in the logs from the ingressgateway that I did hit the svc I wanted but it's giving me a 403 code, permission denied. Anyone has a idea?

**Comment by jovan-absci** [comment]
@N214 did you pass the header as well via `headersToUpstreamOnAllow: ["kubeflow-userid"]`?

**Comment by N214** [comment]
My use case was to enable istio for all applications, kubeflow worked correctly, but I wanted to use istio for other applications that did not need auth. And I was stuck to permission denied.
It was global-deny-all AuthorizationPolicy, so I created another AuthorizationPolicy to whitelist my second service.

Dex-auth-provider is only supposed to trigger with Kubeflow applications, right? hosts: ["kf.example.com"] in your example

**Comment by jovan-absci** [comment]
Yes exactly, we have the same scenario as you. This was written for kubeflow 1.3 and istio that comes bundled with it. Maybe they've added other policies in the meantime.

**Comment by N214** [comment]
Most probably, in the current 1.7 version, there is a global deny all policy and there are several other policies to whitelist each services Kubeflow needed.

Aside from that, not related to this topic, but since you installed kubeflow 1.4, we're at version 1.7. How do you update your kubeflow? Do you update?

**Comment by jovan-absci** [comment]
Yeah, since there isn't much official info out there, we did the following:
1. We stripped Kubeflow down to just notebooks and training operators (left the core services like dex and cert-manager etc). We weren't using the other services much, makes it more lightweight.
2. We switched to a blue/green control plane. We use the green to install new versions from scratch and test them with our other software/images. When it's stable we promote it to be the main control plane and use the old one for future updates.
